[RFC] vim-patch:8.0.0649 and vim-patch:8.0.0650

nimitbhardwaj opened the pull request:
**vim-patch:8.0.0649: when opening a help file the filetype is set several times**
Problem:    When opening a help file the filetype is set several times.
Solution:   When setting the filetype to the same value from a modeline, don't
            trigger FileType autocommands.  Don't set the filetype to "help"
            when it's already set correctly.
**vim-patch 8.0.0650: for extra help files the filetype is set more than once**

Problem:    For extra help files the filetype is set more than once.
Solution:   In *.txt files check that there is no help file modline.

[vim/vim@`904929`](https://github.com/vim/vim/commit/9049298f8d0bbc237b7c666c7ad6cdabe738e8fc) and [vim/vim@`868831`](https://github.com/vim/vim/commit/868831f1224cfc17a4de21fb7386873254592308)

nimitbhardwaj committed:
vim-patch-8.0.0649 and vim-patch-8.0.0650: autocmd open help 2 times

nimitbhardwaj commented:
Closes  #8002

nimitbhardwaj commented:
What is the error

bfredl commented:
The web display of travis logs gets truncated, to see error in oldtest sometimes one has to use "raw log" download. This seems to be the failure:

```
Executed 229 tests
1 FAILED:
Found errors in Test_filetype_detection():
function RunTheTest[35]..Test_filetype_detection[2]..CheckItems line 7: cannot edit "file.text": 
function RunTheTest[35]..Test_filetype_detection[2]..CheckItems line 9: 'with file name: file.text'
function RunTheTest[35]..Test_filetype_detection[2]..CheckItems line 7: cannot edit "README": 
function RunTheTest[35]..Test_filetype_detection[2]..CheckItems line 9: 'with file name: README'
```
you could try to reproduce it locally using `make oldtest`

nimitbhardwaj commented:
OK, thanks, I try to fix the errors

nimitbhardwaj commented:
What can I do here, I think the errors are produced by the runtime/filetype.vim, there are 2 tests which are failed `test_recover.vim` and `test_alot.vim`

nimitbhardwaj commented:
Error is regarding the reading of text files

chrisbra commented on runtime/filetype.vim:
```diff
@@ -2804,8 +2807,13 @@ au BufNewFile,BufRead zsh*,zlog*		call s:StarSetf('zsh')
 " Plain text files, needs to be far down to not override others.  This avoids
 " the "conf" type being used if there is a line starting with '#'.
-au BufNewFile,BufRead *.txt,*.text,README	setf text
+au BufNewFile,BufRead *.text,README self text
```
setf

nimitbhardwaj replied:
yes, I also just noticed it, thanks

chrisbra commented on runtime/filetype.vim:
```diff
@@ -48,6 +48,9 @@ func! s:StarSetf(ft)
 endfunc
 
+" Vim help file
+au BufNewFile,BufRead $VIMRUNTIME/doc/*.txtsetf help
```
missing space

chrisbra commented:
please check your patch. There are a couple of obvious errors in filetype.vim that cause your errors.

nimitbhardwaj commented:
Thank you verymuch

nimitbhardwaj commented:
By correcting it, before there were 2 errors, but now only one 
```text
Test results:


From test_recover.vim:
Found errors in Test_recover_root_dir():
function RunTheTest[35]..Test_recover_root_dir line 11: command did not fail: split Xtest

From test_alot.vim:
Found errors in Test_recover_root_dir():
function RunTheTest[35]..Test_recover_root_dir line 11: command did not fail: split Xtest
TEST FAILURE
```

May be test was somehow expecting line 11 to fail, but it didn't, is it ?

bfredl commented:
Also, adding a functional test that `FileType help` is invoked exactly once when `:help somefile` would probably be good. Look in `test/functional/autocmd/` dir for examples of similar tests.

chrisbra commented:
> May be test was somehow expecting line 11 to fail, but it didn't, is it ?

Are you running as root? The idea of the test is to `:set dir=/` and then when editing a file ( `:sp Xtest`) Vim will throw an error (E303) when trying to create a swapfile in `/`

nimitbhardwaj commented:
yes, I was running as root

nimitbhardwaj commented:
because i installed nvim, I used it so i am working there on it

chrisbra commented:
> yes, I was running as root

Gna, never, never do that!

nimitbhardwaj commented:
okk, so I should run from build/bin/nvim here ?

bfredl commented:
yes. `make oldtest` (and other test commands) will automatically do so.

nimitbhardwaj commented:
Yes, in non root, it worked, thanks now I got that why it did so, and for the test I create a test for the same

bfredl commented on test/functional/autocmd/helpfile_spec.lua:
```diff
@@ -0,0 +1,22 @@
+local helpers = require('test.functional.helpers')(after_each)
```
name the file `filetype_spec.lua`, this way more tests related to `FileType` could be added later.

nimitbhardwaj replied:
ok done

bfredl commented on test/functional/autocmd/helpfile_spec.lua:
```diff
@@ -0,0 +1,22 @@
+describe('autocmd FileType', function()
+    before_each(clear)
+
+    it("triggered by nvim_command('help <cmd>')", function()
```
there is no need to mention the API function, is is incidential, instead something like `'is triggered by :help only once'`

nimitbhardwaj replied:
ok done

nimitbhardwaj commented:
I did it, but still perhaps there will be errors, one error said unable to do functional test, other was about the formating, I corrected some formatting mistakes but, one was not clear to me what that error message mean

nimitbhardwaj commented:
What does this error means 
```text
src/nvim/option.c:3269:  Inner expression should be aligned as opening brace + 1 (+ 2 in case of {). Relevant opening is on line 3268  [whitespace/alignment] [2]
Total errors found: 1
```

bfredl commented on test/functional/autocmd/filetype_spec.lua:
```diff
@@ -0,0 +1,17 @@
+        command("let g:foo = 0")
+        command("autocmd FileType help let g:foo = g:foo + 1")
+        command("help help")
+        assert.same(exp, eval('g:foo'))
```
can just have 1 here, no need for `exp` variable.

nimitbhardwaj replied:
OK, sure

nimitbhardwaj commented:
Done sir

bfredl commented on src/nvim/option.c:
```diff
@@ -2666,6 +2667,7 @@ did_set_string_option (
     } else {
       // load or unload key mapping tables
       errmsg = keymap_init();
+      ft_changed = STRCMP(oldval, *varp) != 0;
```
This doesn't seem to be in the vim patch, nor does FileType autocmd docs mention being fired when `'keymap'` (or `'syntax'` below) is set. What is the motivation for this?

nimitbhardwaj replied:
I was trying if it works, I thought it will work I see the errmsg there so, left it, may be I forgot to remove it, no problem, I remove it

nimitbhardwaj replied:
I was trying if it works, I thought it will work I see the errmsg there so, left it, may be I forgot to remove it, no problem, I remove it

nimitbhardwaj commented:
Done

nimitbhardwaj commented:
Its done here, please if any error, I will be happy to correct it.

bfredl commented on src/nvim/option.c:
```diff
@@ -3174,10 +3175,14 @@ did_set_string_option (
     if (!valid_filetype(*varp)) {
       errmsg = e_invarg;
+    } else {
+      ft_changed = STRCMP(oldval, *varp) != 0;
```
You didn't remove this case which I also mentioned.

bfredl commented:
LGTM after last comment. Please squash and make sure vim-patch:8.0.0649 and vim-patch:8.0.0650 are both included in commit msg, we use this to automatically list merged or NA patches.

nimitbhardwaj commented:
Its done sir, thanks for helping me out

bfredl commented:
Merged, thanks! I think it needs to be `vim-patch:8.0.0649` with colon, so I added it to the merge commit to be sure.

nimitbhardwaj commented:
Thank you very much for merge
